### PR TITLE
TopN optimize for limit 0

### PR DESCRIPTION
**Pull request by kabunchi**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/LimitPushDown.java [hunk 1]
```diff
@@ -168,16 +168,24 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<LimitContext> cont
...
             LimitContext limit = context.get();
+            long count = node.getCount();
+
+            if (limit != null) {
+                count = Math.min(count, limit.getCount());
+            }
+
+            // return empty ValuesNode in case of limit 0
+            if (count == 0) {
+                return new ValuesNode(idAllocator.getNextId(),
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/LimitPushDown.java]: The arguments can all go on the same line. The line's not long enough to warrant splitting it across multiple lines.

**Comment by kabunchi** [comment]
Will add rule and test. 
but it makes we wonder, if there is a rule for evaluatingZeroLimit (missed that) why the code of handling zero limit is needed in LimitPushDown.java::visitLimit ?

**Comment by martint** [comment]
That's a good question. Probable a left-over from when we migrated that logic from the old Visitor-based optimizers that we never cleaned up.

**Comment by martint** [comment]
Oh, I misspoke. That code you're modifying *is* the legacy Visitor-based code. We've been transitioning from that style to the Rule-based optimizers. Because some have no replacement or the legacy ones haven't been removed, they still coexist.

The correct approach here would be to add a rule and leave the legacy code alone.

**Comment by kabunchi** [comment]
opened a new one with rule implementation.

**Pull request closed without merging**